src/check.rs

```diff
@@ -61,16 +122,147 @@ pub fn is_available(url: &Url, ctx: &CheckContext) -> Result<(), CheckError> {
     }
 }
 
+#[cfg(not(test))]
+cached_key_result! {
+    CHECK_FILE: SizedCache<String, HashSet<String>> = SizedCache::with_size(100);
+    Key = { link.without_fragment().to_string() };
+    // `fetch_html` is different depending on whether the link is being
+    // loaded from disk or from the network.
+    fn fragments_from(
+        link: &Link,
+        fetch_html: impl Fn() -> Result<String, CheckError>
+    ) -> Result<HashSet<String>, CheckError> = {
+        fragments_from_inner(fetch_html)
+    }
+}
+
+#[cfg(test)]
+fn fragments_from(
+    _link: &Link,
+    fetch_html: impl Fn() -> Result<String, CheckError>,
+) -> Result<HashSet<String>, CheckError> {
+    fragments_from_inner(fetch_html)
+}
```

Comment:
What's this for? Why is the link being ignored? I'm having a hard time following the code :/

Reply:
Initially cache is keyed by `Link::to_string` which would return different values for same URL with different fragments, e.g `http://example.com/#f1` and `http://example.com/#f2` would be two different keys; and having done the HTTP request for the first and cached it would not help with avoiding the HTTP request for the second. That's why I introduced `Link::without_fragment` which would strip the fragment from a `Url` within `Link::Http`.

Having done that, the tests started to fail as some of them used the same URL (`/` IIRC) to serve different HTML files. So for tests the fetching needs to be without cache.

The code that was in `fragments_for` was moved to `fragments_from_inner` and two variants of `fragments_for` were added, one with and one without caching.

Reply:
I see. I would prefer to fix the tests rather than removing caching altogether - we should test caching too, to make sure it isn't too broad. Can you change the tests to use different URLs instead?

Reply:
Yeah, I've used the test function name as URL for uniqueness.